### PR TITLE
[DW-4185] Add cloudwatch metric filters and alarms

### PR DESCRIPTION
**Pull request by liamraemclauchlan**
* Create cloudwatch log metric filter
* Create alarm when metric threshold is hit
* Send alert to SNS topic
* Triggers lambda to send slack notification

module documentation - https://registry.terraform.io/modules/dwp/metric-filter-alarm/aws/1.1.5
examples - https://github.ucds.io/dip/aws-ingestion/blob/master/kafka_to_hbase_monitoring.tf#L59...L92

### CODE REVIEW

File: terraform/deploy/app/orchestration_service.tf [hunk 1]
```diff
@@ -203,3 +203,38 @@ module "jupyter_s3_storage" {
...
   vpc_id         = data.terraform_remote_state.aws_analytical_env_infra.outputs.vpc.aws_vpc
 }
+
+#
+## ---------------------------------------------------------------------------------------------------------------------
+## Cloudwatch Log Metrics/Alarms
+## ---------------------------------------------------------------------------------------------------------------------
+
+module "orchestration_service_failed_to_destroy_alarm" {
+  source = "dwp/metric-filter-alarm/aws"
```
Comment (dan-hill2802) [inline terraform/deploy/app/orchestration_service.tf:231]: can you add a version please

File: terraform/deploy/app/orchestration_service.tf [hunk 2]
```diff
@@ -203,3 +203,38 @@ module "jupyter_s3_storage" {
...
+  alarm_action_arns   = [data.terraform_remote_state.security-tools.outputs.sns_topic_london_monitoring.arn]
+  evaluation_periods  = "1"
+  period              = 60
+  threshold           = 0
+  statistic           = "Sum"
+  comparison_operator = "GreaterThanThreshold"
+}
+
+module "guacamole_invalid_user_desktop_alarm" {
+  source = "dwp/metric-filter-alarm/aws"
```
Comment (dan-hill2802) [inline terraform/deploy/app/orchestration_service.tf:246]: can you add a version please

File: terraform/deploy/app/orchestration_service.tf [hunk 3]
```diff
@@ -203,3 +203,38 @@ module "jupyter_s3_storage" {
...
+## Cloudwatch Log Metrics/Alarms
+## ---------------------------------------------------------------------------------------------------------------------
+
+module "orchestration_service_failed_to_destroy_alarm" {
+  source = "dwp/metric-filter-alarm/aws"
+
+  log_group_name      = module.ecs-fargate-task-definition.log_group
+  metric_namespace    = "/app/${var.name_prefix}"
+  pattern             = "{ $.message = \"Failed to destroy*\"}"
+  alarm_name          = "FailedToDestroy"
```
Comment (dan-hill2802) [inline terraform/deploy/app/orchestration_service.tf]:
```suggestion
  alarm_name          = "Orchestration Service Failed to Destroy"
```

File: terraform/deploy/app/orchestration_service.tf [hunk 4]
```diff
@@ -203,3 +203,38 @@ module "jupyter_s3_storage" {
...
+  comparison_operator = "GreaterThanThreshold"
+}
+
+module "guacamole_invalid_user_desktop_alarm" {
+  source = "dwp/metric-filter-alarm/aws"
+
+  log_group_name      = module.ecs-user-host.outputs.user_container_log_group
+  metric_namespace    = "/app/${var.name_prefix}-user-host-guacamole"
+  pattern             = "Cognito user tried to access desktop for" # No consistent JSON logging for these container_logs, so pattern is plaintext search
+  alarm_name          = "InvalidUserDesktopAccess"
```
Comment (dan-hill2802) [inline terraform/deploy/app/orchestration_service.tf]:
```suggestion
  alarm_name          = "Orchestration Service Invalid User Desktop Access"
```

File: terraform/modules/fargate-task-definition/outputs.tf [hunk 5]
```diff
@@ -21,3 +21,8 @@ output "container_name" {
   description = "Name of the container"
   value       = var.container_name
 }
+
+output "log_group" {
+  description = "Name of the log group for OS"
+  value       = aws_cloudwatch_log_group.lamda_logs.name
```
Comment (dan-hill2802) [inline terraform/modules/fargate-task-definition/outputs.tf:27]:
```suggestion
  description = "Log group for OS"
  value       = {
      name = aws_cloudwatch_log_group.lamda_logs.name
      arn = aws_cloudwatch_log_group.lamda_logs.arn
```
That will need `terraform fmt` applying

File: terraform/deploy/app/orchestration_service.tf [hunk 6]
```diff
@@ -221,3 +221,36 @@ module "cleanup_lambda" {
...
+module "invalid_user_desktop_alarm" {
+  source = "dwp/metric-filter-alarm/aws"
+
+  log_group_name   = module.ecs-user-host.outputs.user_container_log_group
+  metric_namespace = "/app/guacamole"
+  pattern          = "Cognito user tried to access desktop for"
+  # No consistent JSON logging for these container_logs, so pattern is plaintext search
+  alarm_name = "InvalidUserDesktopAccess"
+  alarm_action_arns = [
+  data.terraform_remote_state.security-tools.outputs.sns_topic_london_monitoring.arn]
```
Comment (dan-hill2802) [inline terraform/deploy/app/orchestration_service.tf:254]:
```suggestion
  alarm_action_arns = [data.terraform_remote_state.security-tools.outputs.sns_topic_london_monitoring.arn]
```

File: terraform/deploy/app/orchestration_service.tf (hunk 1 above)
Comment (liamraemclauchlan) [reply terraform/deploy/app/orchestration_service.tf]: Done

File: terraform/deploy/app/orchestration_service.tf (hunk 2 above)
Comment (liamraemclauchlan) [reply terraform/deploy/app/orchestration_service.tf]: Done

File: terraform/deploy/app/orchestration_service.tf (hunk 3 above)
Comment (liamraemclauchlan) [reply terraform/deploy/app/orchestration_service.tf]: :+1:

File: terraform/modules/fargate-task-definition/outputs.tf [hunk 7]
```diff
@@ -21,3 +21,11 @@ output "container_name" {
   description = "Name of the container"
   value       = var.container_name
 }
+
+output "log_group" {
+  description = "Name of the log group for OS"
```
Comment (dan-hill2802) [inline terraform/modules/fargate-task-definition/outputs.tf:26]: it's not really the name, it's multiple attributes

Comment (liamraemclauchlan) [reply terraform/modules/fargate-task-definition/outputs.tf:26]: touché

**Pull request merged**